### PR TITLE
8312433: HttpClient request fails due to connection being considered idle and closed

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this change which proposes to fix the issue noted in https://bugs.openjdk.org/browse/JDK-8312433?

In JDK 20, we introduced a way to identify and close idle HTTP2 connections in the HttpClient implementation through https://bugs.openjdk.org/browse/JDK-8288717. Whenever a HTTP2 stream gets closed and if there are no other streams on the connection, we create a idle timeout event that times out after a (configurable) duration. After the event has been registered, if any more streams get opened on that connection, then the event is cancelled. If no streams are created during that period, then the event fires and we close the connection.
We have a race condition in this implementation, as noticed in https://bugs.openjdk.org/browse/JDK-8312433. When a HTTP2 connection is created, we pool that connection. Any subsequent request against the same target server will reuse this pooled connection if it's still open. When user code issues a request, we checkout a connection from this pool (if relevant) and then use that connection to create a HTTP2 stream and issue the request. This sequence has a small window of time, where a pooled connection which has a idle timeout event scheduled (because it had no active streams) is handed out of the pool and before it can create a new HTTP2 stream, gets closed (concurrently by the idle connection timeout event). This results in the HTTP2 stream creation to fail with an exception that effectively fails the user initiated request.

The commit in this PR addresses it by introducing an internal method `tryReserveForPoolCheckout()` on the `Http2Connection` which is responsible for co-ordinating with the idle connection management to ensure that if this connection does get handed out of the pool, then the idle connection event doesn't end up closing it.

A new jtreg test has been added to test this issue. This test consistently reproduces the issue without this fix and passes with this fix. Ideally, a new test method could have been added to the existing `test/jdk/java/net/httpclient/http2/IdleConnectionTimeoutTest.java` but that test has multiple `@run` with different timeout values ranging all the way upto 30 seconds and it focuses more on the timeout value parsing and honouring that value. Introducing this test method in that existing test would end up testing this against all those values for no reason, plus would also end up increasing the duration of that test. So I decided to introduce a new test which uses a much lower timeout value.

tier1, tier2 and tier3 testing has passed with this change and multiple test runs of `test/jdk/java/net/httpclient` too has passed with this change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312433](https://bugs.openjdk.org/browse/JDK-8312433): HttpClient request fails due to connection being considered idle and closed (**Bug** - P4)

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15012/head:pull/15012` \
`$ git checkout pull/15012`

Update a local copy of the PR: \
`$ git checkout pull/15012` \
`$ git pull https://git.openjdk.org/jdk.git pull/15012/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15012`

View PR using the GUI difftool: \
`$ git pr show -t 15012`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15012.diff">https://git.openjdk.org/jdk/pull/15012.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15012#issuecomment-1649414028)

### CODE REVIEW

File: src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java [hunk 1]
```diff
@@ -131,7 +131,7 @@ class Http2Connection  {
 
     private static final int MAX_CLIENT_STREAM_ID = Integer.MAX_VALUE; // 2147483647
     private static final int MAX_SERVER_STREAM_ID = Integer.MAX_VALUE - 1; // 2147483646
-    private IdleConnectionTimeoutEvent idleConnectionTimeoutEvent;  // may be null
+    private volatile IdleConnectionTimeoutEvent idleConnectionTimeoutEvent;  // may be null
```
Comment (djelinski) [inline src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:134]: What was the reason for this change? `idleConnectionTimerEvent` is always protected by `stateLock`.

**Review by djelinski — COMMENTED** [review]
Hi Jaikiran, thanks for taking care of this. Changes look good, except for `volatile` usage that looks unnecessary.

File: src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java (hunk 1 above)
Comment (jaikiran) [reply src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java:134]: Hello Daniel, you are right. I missed cleaning up that part from my early experiments. I'll fix that and do the necessary changes shortly.

**Comment by jaikiran** [comment]
I've now updated the PR to remove the update which used `volatile` and cleaned up its usage a bit. Tests continue to pass with this change.

**Comment by jaikiran** [comment]
Thank you Daniel for the review. CI tests passed with these changes. I'll go ahead with integrating this.

**Pull request closed without merging**